Wikipedia/View Controllers/ShareCard/WMFShareOptionsController.m

```diff
@@ -0,0 +1,346 @@
+
+- (void)presentActivityViewControllerWithImage:(nullable UIImage*)image title:(NSString*)title {
+   
+    NSString* parameter = image ? @"wprov=sfii1" : @"wprov=sfti1";
+
+    NSURL* url = [[NSURL alloc] initWithString:[NSString stringWithFormat:@"%@?%@",
+                                                self.article.title.desktopURL.absoluteString,
+                                                parameter]];
+
+    NSMutableArray* activityItems = @[title, url].mutableCopy;
+    if (image) {
+        [activityItems addObject:image];
+    }
+    
+    UIActivityViewController* shareActivityVC =
+    [[UIActivityViewController alloc] initWithActivityItems:activityItems
+                                      applicationActivities:@[] /*shareMenuSavePageActivity*/ ];
+    
+    shareActivityVC.excludedActivityTypes = @[
+                                              UIActivityTypePrint,
+                                              UIActivityTypeAssignToContact,
+                                              UIActivityTypeAirDrop,
+                                              UIActivityTypeAddToReadingList
+                                              ];;
+    
+    [shareActivityVC setCompletionWithItemsHandler:^(NSString * __nullable activityType, BOOL completed, NSArray * __nullable returnedItems, NSError * __nullable activityError){
+        if (completed) {
+            [self.shareFunnel logShareSucceededWithShareMethod:activityType];
+        } else {
+            [self.shareFunnel logShareFailedWithShareMethod:activityType];
+        }
+    }];
+    
+    if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPhone) {
+        [self.containerViewController presentViewController:shareActivityVC animated:YES completion:nil];
+    } else {
+        
+        self.popover = [[UIPopoverController alloc] initWithContentViewController:shareActivityVC];
+        self.popover.delegate = self;
+        
+        if(self.originButtonItem){
+
+            [self.popover presentPopoverFromBarButtonItem:self.originButtonItem
+                                 permittedArrowDirections:UIPopoverArrowDirectionAny
+                                                 animated:YES];
+        }else{
+            
+            CGRect frame = self.originView.frame;
+            if(CGRectIsNull(frame)){
+                frame  = self.containerViewController.view.frame;
+            }else{
+                frame = [self.containerViewController.view convertRect:frame fromView:self.originView.superview];
+            }
+            
+            [self.popover presentPopoverFromRect:frame
+                                          inView:self.containerViewController.view
+                        permittedArrowDirections:UIPopoverArrowDirectionAny
+                                        animated:YES];
+        }
+    }
+    
+    [self cleanup];
+}
+
+- (void)popoverControllerDidDismissPopover:(UIPopoverController *)popoverController{
+    self.popover = nil;
```

Comment:
oh, i think i got it. `cleanup` removes the share options view, which doesn't need to happen for a popover?